### PR TITLE
feat: force email validation if attendee exists in database

### LINKED ISSUE #15252: [CAL-3829] force email validation if attendee exists in database
- when adding an attendee email, check if cal.com user with that email exists
- if no, make booking as normal
- if yes, require email validation (change confirmation button to “verify”

- unless email owner is logged in (check session). if logged in, no verification needed

<sub>[CAL-3829](https://linear.app/calcom/issue/CAL-3829/force-email-validation-if-attendee-exists-in-database)</sub>

### PR DESCRIPTION
**Pull request by Amit91848**
## What does this PR do?

- Fixes #15252 
- Fixes CAL-3829

Logged in with teampro and teampro has s@s.com as secondary email.

https://github.com/calcom/cal.com/assets/74371312/7614bc44-27e1-462a-8ab4-6db6d92e634c

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [ ] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

### CODE REVIEW

**Review by PeerRich — COMMENTED** [review]
awesome work! does it take all verified emails of an account?

what about + variantions i.e. peer+2@cal.com when peer@cal.com is verified 🤔

**Comment by Amit91848** [comment]
> awesome work! does it take all verified emails of an account?

Yes, it also checks for secondary verified emails of the account.

> what about + variantions i.e. [peer+2@cal.com](mailto:peer+2@cal.com) when [peer@cal.com](mailto:peer@cal.com) is verified 🤔

No, currently it checks for the email as is. Do you want me to take variations into consideration too?

**Comment by emrysal** [comment]
Hey @Amit91848 - if you could add the + variations that would be very appreciated. The feature is easily 'circumventable' if this + is not handled properly which makes it less useful. Let's fix it properly :)

File: packages/trpc/server/routers/publicViewer/userWithEmail.handler.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import type { TUserWithEmailInputSchema } from "./userWithEmail.schema";
+
+type UserWithEmailOptions = {
+  input: TUserWithEmailInputSchema;
+};
+
+export const userWithEmailHandler = async ({ input }: UserWithEmailOptions) => {
+  const { userEmail, email } = input;
+
+  const user = await prisma.user.findFirst({
```
Comment (keithwillcode) [inline packages/trpc/server/routers/publicViewer/userWithEmail.handler.ts:12]: Don't we need to check `emailVerified` on the `user` table as well?

**Comment by keithwillcode** [comment]
In the Loom, this only addresses the main email field. Does it also prevent emails from going into the guest list?

**Comment by keithwillcode** [comment]
> In the Loom, this only addresses the main email field. Does it also prevent emails from going into the guest list?

Perhaps we remove guests from the list as I did in a PoC #15255 if the guest email isn't verified. 

In either case, we need to take the email used and see if it's a variation as mentioned above.

File: packages/features/bookings/Booker/components/hooks/useVerifyEmail.ts [hunk 2]
```diff
@@ -32,6 +36,11 @@ export const useVerifyEmail = ({
     },
   });
 
+  const { data: userWithEmail } = trpc.viewer.public.userWithEmail.useQuery({
```
Comment (Udit-takkar) [inline packages/features/bookings/Booker/components/hooks/useVerifyEmail.ts:39]: NIT: change the name of variable `userWithEmail`.

File: packages/trpc/server/routers/publicViewer/userWithEmail.handler.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    where: {
+      OR: [
+        { email },
```
Comment (Udit-takkar) [inline packages/trpc/server/routers/publicViewer/userWithEmail.handler.ts:15]: Need to check emailVerified field here

File: packages/features/bookings/Booker/components/hooks/useVerifyEmail.ts (hunk 2 above)
Comment (keithwillcode) [reply packages/features/bookings/Booker/components/hooks/useVerifyEmail.ts:39]: Yeah agreed. Let's change that. It doesn't really tell you what it is accomplishing.

File: packages/trpc/server/routers/publicViewer/checkIfUserExistWithEmail.handler.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  input: TUserWithEmailInputSchema;
+};
+
+// Function to extract base email
+const extractBaseEmail = (email: string): string => {
+  const [localPart, domain] = email.split("@");
+  const baseLocalPart = localPart.split("+")[0];
+  return `${baseLocalPart}@${domain}`;
+};
+
```
Comment (Udit-takkar) [inline packages/trpc/server/routers/publicViewer/checkIfUserExistWithEmail.handler.ts:15]: this would check for pro+2@example.com

Comment (zomars) [reply packages/trpc/server/routers/publicViewer/checkIfUserExistWithEmail.handler.ts:15]: Are we supposed to not allow these?

**Comment by keithwillcode** [comment]
@PeerRich This PR is not the one to go with imho because this still allows spammers to fire off OTP email verification spam to email addresses. At least in the short-term, we should ship #15255 and iterate from there.

**Comment by keithwillcode** [comment]
This functionality has been merged with #15255

**Pull request closed without merging**